### PR TITLE
ci: :construction_worker: changed to building imagick from source

### LINKED ISSUE #2425: Error on docker build on the pecl build of imagick library.
### Prerequisites

- [X] I've [searched](https://github.com/accessibility-exchange/platform/issues) for existing issues

### Describe the bug

```
    - 'Imagick::newImage(): Parameter format has null default, but is not nullable'
    - 'make: *** [Makefile:191: /tmp/pear/temp/imagick/Imagick_arginfo.h] Error 1'
    - 'ERROR: `make INSTALL_ROOT="/tmp/pear/temp/pear-build-defaultuserj9kc76m1e7redzixmd2/install-imagick-3.7.0" install'' failed'
```  
Found issue on the libary github.  
https://github.com/Imagick/imagick/issues/698

### PR DESCRIPTION
**Pull request by marvinroman**
Updated Dockerfile that builds the main application image to build the imagick library from source instead of using `pecl` since their version is not yet fully ready for php v8.4.

### CODE REVIEW

File: .kube/app/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,7 @@
 FROM php:8.4-fpm
 
 ENV PHP_VERSION 8.4
+ENV IMAGICK_VERSION 3.7.0 # This should be bumped to the latest version and
```
Comment (jobara) [inline .kube/app/Dockerfile]: seems like part of the comment is missing

File: .kube/app/Dockerfile [hunk 2]
```diff
@@ -36,9 +37,29 @@ RUN apt-get install -y \
...
 RUN docker-php-ext-install -j$(nproc) gd pdo_mysql mbstring exif pcntl bcmath bz2 zip soap intl opcache
 
-RUN apt-get install -y libmagickwand-dev --no-install-recommends
+RUN apt-get install -y \
+    libmagickwand-dev \
+    libmagickcore-dev \
+    --no-install-recommends
 
-RUN pecl install imagick
+# START BUILDING IMAGICK FROM SOURCE - this can be removed once pecl fixes issues installing with php8.4
```
Comment (jobara) [inline .kube/app/Dockerfile:44]: Can you please leave a link to https://github.com/accessibility-exchange/platform/issues/2425 in the comments. It'll just make it easier to understand when scanning the code later.

File: .kube/app/Dockerfile [hunk 3]
```diff
@@ -48,6 +69,11 @@ RUN pecl install -o -f redis && \
     rm -rf /tmp/pear && \
     docker-php-ext-enable redis
 
+# - this can be removed once pecl fixes issues installing with php8.4
+RUN apt-get remove -y \
+    $PHPIZE_DEPS \
+    libtool
```
Comment (jobara) [inline .kube/app/Dockerfile]: Is it possible to do the remove in the same block as the install? I think it's more likely this will get missed when reverting later.

**Review by jobara — CHANGES_REQUESTED** [review]
What's the best way to test these changes before merging?

File: .kube/app/Dockerfile (hunk 1 above)
Comment (marvinroman) [reply .kube/app/Dockerfile]: updated

File: .kube/app/Dockerfile (hunk 2 above)
Comment (marvinroman) [reply .kube/app/Dockerfile:44]: added

File: .kube/app/Dockerfile (hunk 3 above)
Comment (marvinroman) [reply .kube/app/Dockerfile]: It needs to be there or the following pecl build commands fail. I thought those commands ensures build dependencies were installed if necessary when pecl runs, but it failed when I removed them prior to running those.

Comment (marvinroman) [reply .kube/app/Dockerfile]: Migrated to TODO comments which should be recognized by IDE's that implement finding the tags for pending items that are pending changes.

File: .kube/app/Dockerfile [hunk 4]
```diff
@@ -1,6 +1,8 @@
 FROM php:8.4-fpm
 
 ENV PHP_VERSION 8.4
+# TODO - IMAGICK_VERSION can be removed once pecl fixes issues installing with php8.4
+ENV IMAGICK_VERSION 3.7.0 # This should be bumped to the latest version, especially when PHP version is bumped and can be removed when https://github.com/accessibility-exchange/platform/issues/2425 is resolved
```
Comment (jobara) [inline .kube/app/Dockerfile]: I guess this link is a bit odd because this PR will close that issue. Maybe reword a bit here so that it indicates that 2425 is what this work around is for an can be removed when the upstream issue in imagick is resolved (and link to an imagick issue).

**Review by jobara — CHANGES_REQUESTED** [review]

File: .kube/app/Dockerfile (hunk 4 above)
Comment (marvinroman) [reply .kube/app/Dockerfile]: Changed all the TODO comments to reference issue https://github.com/Imagick/imagick/issues/698.

File: .kube/app/Dockerfile [hunk 5]
```diff
@@ -36,9 +38,30 @@ RUN apt-get install -y \
...
-RUN pecl install imagick
+# START BUILDING IMAGICK FROM SOURCE - this can be removed once pecl fixes issues installing with php8.4
+# TODO - this can be removed when https://github.com/Imagick/imagick/issues/698 is resolved
+RUN apt-get install -y \
+    $PHPIZE_DEPS \
+    libtool
+
+RUN mkdir -p /tmp/imagick
+WORKDIR /tmp/imagick
+RUN curl -L -o /tmp/imagick.tar.gz https://github.com/Imagick/imagick/archive/tags/${IMAGICK_VERSION}.tar.gz
```
Comment (jobara) [inline .kube/app/Dockerfile]: When I tried to build the container I got the error 

```
failed to solve: process "/bin/sh -c curl -L -o /tmp/imagick.tar.gz https://github.com/Imagick/imagick/archive/tags/${IMAGICK_VERSION}.tar.gz" did not complete successfully: exit code: 6
```

Comment (marvinroman) [reply .kube/app/Dockerfile]: When abstracting version to an environment variable it broke the curl command reverted to static url.

**Comment by jobara** [comment]
@marvinroman I'm having trouble testing this with the local docker compose. Every time I get to a spot that needs to start mysql it fails. It says it can't properly start it or that it's unhealthy. Have you run into an issue like this?

**Comment by jobara** [comment]
@marvinroman I've tried a few more times after updating my docker tools and it is still failing to start mysql. I asked @greatislander to also try this PR but he was unable to run the local docker compose based environment due to issues running it on an Apple Silicon based Mac.

**Comment by marvinroman** [comment]
> @marvinroman I've tried a few more times after updating my docker tools and it is still failing to start mysql. I asked @greatislander to also try this PR but he was unable to run the local docker compose based environment due to issues running it on an Apple Silicon based Mac.

It's not just an issue with Apple and I was able to resolve the issue by eliminating the custom MySQL configuration but this configuration is the same as what gets deployed on the Kube. I ran out of time last week to finish the full update and to find out which configuration variable it causing the issues. I will submit changes this week to resolve.

**Comment by marvinroman** [comment]
@jobara The reported errors in deployment have been fixed in branch **infrastructure-local-environment-update-nix** you will want to merge both these branches locally to be able to fully build and deploy.  

To complete the updates I would do the following after downloading the updates.  
* Backing up the `.env` file and letting it be recreated.  
* Rebuild the laravel container using the following alias `dcbp`.  
* Remove the mysql volume.

**Comment by jobara** [comment]
@marvinroman I got things built using nix, but even when combining the PRs I still got a failure with mysql ( `dependency failed to start: container platform.mysql exited (1)` ) when running docker local without nix. This happens in the step for creating the app key. Also I had to `WWWUSER` and `WWWGROUP` info before I could get this step to get even this far.

**Comment by marvinroman** [comment]
> @marvinroman I got things built using nix, but even when combining the PRs I still got a failure with mysql ( `dependency failed to start: container platform.mysql exited (1)` ) when running docker local without nix. This happens in the step for creating the app key. Also I had to `WWWUSER` and `WWWGROUP` info before I could get this step to get even this far.

Can you give me the contents of `docker logs platform.mysql` ? Did you remove the volume for the mysql container? If you didn't you will want to make sure to copy the database credentials from the backed up `.env` file before you let `nix-shell` recreate it. But personally I would just delete the volume so it can be started fresh with new credentials.

You should no longer need `WWWGROUP`.

Did you delete the `.env` file and then ran `nix-shell` and when it was re-created the `WWWUSER` was empty?

**Comment by jobara** [comment]
> > @marvinroman I got things built using nix, but even when combining the PRs I still got a failure with mysql ( `dependency failed to start: container platform.mysql exited (1)` ) when running docker local without nix. This happens in the step for creating the app key. Also I had to `WWWUSER` and `WWWGROUP` info before I could get this step to get even this far.
> 
> Can you give me the contents of `docker logs platform.mysql` ? Did you remove the volume for the mysql container? If you didn't you will want to make sure to copy the database credentials from the backed up `.env` file before you let `nix-shell` recreate it. But personally I would just delete the volume so it can be started fresh with new credentials.
> 
> You should no longer need `WWWGROUP`.
> 
> Did you delete the `.env` file and then ran `nix-shell` and when it was re-created the `WWWUSER` was empty?

@marvinroman The issue I'm mentioning here is when running without nix ( I did get something to build using nix and have merged #2429 ) and using just the instructions from "Local development setup using docker compose:". I still get the issues I mentioned above with the ID & Group needing to happen earlier and mysql failing to start.

**Comment by jobara** [comment]
@marvinroman here are the logs from `docker logs platform.mysql`. Looks like it's an issue with the mysql root password.

```log
mysql 19:00:12.12 INFO  ==> 
mysql 19:00:12.12 INFO  ==> Welcome to the Bitnami mysql container
mysql 19:00:12.12 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
mysql 19:00:12.13 INFO  ==> Did you know there are enterprise versions of the Bitnami catalog? For enhanced secure software supply chain features, unlimited pulls from Docker, LTS support, or application customization, see Bitnami Premium or Tanzu Application Catalog. See https://www.arrow.com/globalecs/na/vendors/bitnami/ for more information.
mysql 19:00:12.13 INFO  ==> 
mysql 19:00:12.13 INFO  ==> ** Starting MySQL setup **
mysql 19:00:12.15 INFO  ==> Validating settings in MYSQL_*/MARIADB_* env vars
mysql 19:00:12.16 ERROR ==> The MYSQL_ROOT_PASSWORD environment variable is empty or not set. Set the environment variable ALLOW_EMPTY_PASSWORD=yes to allow the container to be started with blank passwords. This is recommended only for development.
mysql 19:00:12.16 ERROR ==> root user is already created in the database and you can't use it as username for user creation.
mysql 19:11:47.77 INFO  ==> 
mysql 19:11:47.78 INFO  ==> Welcome to the Bitnami mysql container
mysql 19:11:47.78 INFO  ==> Subscribe to project updates by watching https://github.com/bitnami/containers
mysql 19:11:47.78 INFO  ==> Did you know there are enterprise versions of the Bitnami catalog? For enhanced secure software supply chain features, unlimited pulls from Docker, LTS support, or application customization, see Bitnami Premium or Tanzu Application Catalog. See https://www.arrow.com/globalecs/na/vendors/bitnami/ for more information.
mysql 19:11:47.78 INFO  ==> 
mysql 19:11:47.79 INFO  ==> ** Starting MySQL setup **
mysql 19:11:47.81 INFO  ==> Validating settings in MYSQL_*/MARIADB_* env vars
mysql 19:11:47.82 ERROR ==> The MYSQL_ROOT_PASSWORD environment variable is empty or not set. Set the environment variable ALLOW_EMPTY_PASSWORD=yes to allow the container to be started with blank passwords. This is recommended only for development.
mysql 19:11:47.82 ERROR ==> root user is already created in the database and you can't use it as username for user creation.
```

**Comment by jobara** [comment]
@marvinroman just a heads up that the laravel-vite-plugin was recently updated with this change ( https://github.com/laravel/vite-plugin/pull/318 ). I don't know if it will affect any of the local docker or nix setups, but just letting you know in case.

**Comment by marvinroman** [comment]
> @marvinroman just a heads up that the laravel-vite-plugin was recently updated with this change ( [laravel/vite-plugin#318](https://github.com/laravel/vite-plugin/pull/318) ). I don't know if it will affect any of the local docker or nix setups, but just letting you know in case.

It shouldn't.

**Comment by marvinroman** [comment]
@jobara I would rebase or update from **dev** both those issues should be addressed by the new nix environment.

**Comment by jobara** [comment]
> @jobara I would rebase or update from **dev** both those issues should be addressed by the new nix environment.

@marvinroman Yes I've been testing by first creating a test branch off of dev and merging this PR's branch into it. The problem here though is that there is still documentation, and I suppose artifacts in the repo, for running "Local development setup using docker compose". If this method for local development is no longer supported all those parts should be removed.

**Comment by marvinroman** [comment]
> > @jobara I would rebase or update from **dev** both those issues should be addressed by the new nix environment.
> 
> @marvinroman Yes I've been testing by first creating a test branch off of dev and merging this PR's branch into it. The problem here though is that there is still documentation, and I suppose artifacts in the repo, for running "Local development setup using docker compose". If this method for local development is no longer supported all those parts should be removed.

@jobara That is true. Initially I wanted to keep both to not make Nix a requirement. But I think some of it broke with the new environment and it's better to not try and maintain both options. That section is now removed.

**Review by jobara — APPROVED** [review]
